lib/tcpip/network_wrapper.cc

```diff
@@ -444,17 +456,19 @@ int network_socket_close(Timeout *t, SObj mallocCapability, SObj sealedSocket)
 		  {
 			  return -ETIMEDOUT;
 		  }
-		  socket->socketLock.upgrade_for_destruction();
-		  // Drop the caller's claim on the socket.
-		  if (heap_free(mallocCapability, socket->socket) != 0)
+		  // Since we free the socket and the token at the end after
+		  // terminating the socket, ensure that the frees won't fail
+		  if (heap_can_free(mallocCapability, socket->socket) != 0 ||
+		      token_obj_can_destroy(
+		        mallocCapability, socket_key(), sealedSocket) != 0)
 		  {
-			  Debug::log("Failed to free socket");
-			  // Release the lock so that we don't leak it.
-			  socket->socketLock.unlock();
-			  return -EINVAL;
+			  return -EAGAIN; // TODO EINVAL?
```

Comment:
Should be `EINVAL` or `EPERM`: it's definitely not `EAGAIN`: retrying with the wrong thing won't make it right.